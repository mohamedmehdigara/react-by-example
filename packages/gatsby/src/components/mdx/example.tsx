import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
