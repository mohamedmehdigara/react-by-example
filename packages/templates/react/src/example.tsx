import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
