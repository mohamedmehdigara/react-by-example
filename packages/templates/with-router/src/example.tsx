import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
