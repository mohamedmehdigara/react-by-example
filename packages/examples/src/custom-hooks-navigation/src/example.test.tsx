import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { FC, useMemo } from 'react'
import { Tabs } from './example'

describe('Tabs', () => {
  let component: RenderResult

  const Test: FC = () => {
    const tabs = useMemo(
      () => [
        { id: 'hello', title: 'Hello', content: <aside>hello</aside> },
        { id: 'world', title: 'World', content: <aside>world</aside> },
        { id: 'tabs', title: 'Tabs', content: <aside>tabs</aside> },
        { id: 'example', title: 'Example', content: <aside>example</aside> },
      ],
      [],
    )
    return <Tabs defaultTabId="tabs" tabs={tabs} />
  }

  beforeEach(() => {
    component = render(<Test />)
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should activate default tab', () => {
    expect(component.getByRole('complementary')).toHaveTextContent('tabs')
  })

  describe('when switching tabs', () => {
    beforeEach(() => {
      userEvent.click(component.getByText('Hello'))
    })

    it('should activate tab', () => {
      expect(component.getByRole('complementary')).toHaveTextContent('hello')
    })
  })

  describe('when using keyboard navigation', () => {
    let b1: HTMLElement
    let b2: HTMLElement
    let b3: HTMLElement
    let b4: HTMLElement

    beforeEach(() => {
      ;[b1, b2, b3, b4] = component.getAllByRole('button')
    })

    describe('when pressing left', () => {
      beforeEach(() => {
        fireEvent.keyDown(b1, { key: 'ArrowLeft' })
      })

      it('should focus tab', () => {
        expect(b4).toHaveFocus()
      })
    })

    describe('when pressing right', () => {
      beforeEach(() => {
        fireEvent.keyDown(b2, { key: 'ArrowRight' })
      })

      it('should focus tab', () => {
        expect(b3).toHaveFocus()
      })
    })
  })
})
