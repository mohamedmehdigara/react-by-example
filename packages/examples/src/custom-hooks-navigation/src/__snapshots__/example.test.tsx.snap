// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should snapshot 1`] = `
<div>
  <div>
    <div>
      <button
        class="button"
        type="button"
      >
        
        Hello
      </button>
      <button
        class="button"
        type="button"
      >
        
        World
      </button>
      <button
        class="button"
        type="button"
      >
        &gt; 
        Tabs
      </button>
      <button
        class="button"
        type="button"
      >
        
        Example
      </button>
    </div>
    <div>
      <aside>
        tabs
      </aside>
    </div>
  </div>
</div>
`;
